tcms/core/ajax.py

```diff
@@ -240,6 +198,57 @@ def remove(self):
     return render(request, template_name, context_data)
 
 
+class _TagObjects(object):
+    """ Used for getting the chosen object(TestPlan, TestCase or TestRun) from the database """
+
+    def __init__(self, request):
+        """
+        :param request: An HTTP GET request, containing the private key
+                        and the type of object to be selected
+        :type request: HttpRequest
+        """
+        for obj in ['plan', 'case', 'run']:
+            if request.GET.get(obj):
+                self.object = obj
+                self.object_pk = request.GET.get(obj)
+                break
+
+    def get(self):
+        func = getattr(self, self.object)
+        return func()
+
+    def plan(self):
+        return 'management/get_tag.html', TestPlan.objects.get(pk=self.object_pk)
+
+    def case(self):
+        return 'management/get_tag.html', TestCase.objects.get(pk=self.object_pk)
+
+    def run(self):
+        return 'run/get_tag.html', TestRun.objects.get(pk=self.object_pk)
+
+
+class _TagActions(object):
+    """ Used for performing the 'add' and 'remove' actions on a given tag """
+
+    def __init__(self, obj, tag_name):
+        """
+        :param obj: the object for which the tag actions would be performed
+        :type obj: either a TestPlan, TestCase ot TestRun
```

Comment:
See the XML-RPC api docs on how to annotate class names so that Sphinx will automatically link to them.

Reply:
Sphinx would not allow having three class names as type of one object. I suppose we leave it as it is.

Reply:
How do you know Sphinx doesn't allow that since documentation for classes and methods starting with underscore isn't shown automatically ? 

The standard reference syntax is, e.g. :class:`tcms.testplan.models.models.TestPlan` which allows Sphynx to put a link directly to its documentation. However this doesn't seem to work for type annotations so you can keep the short name syntax. 

Fix the typo though.